kong/plugins/ai-proxy/handler.lua

```diff
@@ -0,0 +1,183 @@
+local _M = {}
+
+-- imports
+local ai_shared = require("kong.plugins.ai-proxy.drivers.shared")
+local fmt = string.format
+local kong_utils = require("kong.tools.utils")
+local kong_meta = require "kong.meta"
+--
+
+_M.PRIORITY = 769
+_M.VERSION = kong_meta.version
+
+local INTERNAL_SERVER_ERROR_TEMPLATE = '{"error":{"message":"%s"}}'
+
+local function do_bad_request(msg)
+  kong.log.warn(msg)
+  kong.response.exit(400, { message = msg, error = true })
+end
+
+local function do_internal_server_error(msg)
+  kong.log.warn(msg)
+  kong.response.exit(500, { message = msg, error = true })
+end
+
+local function apply_transformations(route_type, request_table, transformation_instructions)
+  -- this is mostly a placeholder function for now;
+  -- it just receives a replacement "template" from a previous
+  -- AI-plugin, and applies it over the top of the existing request body
+  local this = {}
+   
+  -- use template engine
+  if route_type == "llm/v1/completions" then
+    kong.log.debug("template is llm/v1/completions type")
+    this = {
+      prompt = transformation_instructions[1],
+    }
+  elseif route_type == "llm/v1/chat" then
+    kong.log.debug("template is llm/v1/chat type")
+    this = {
+      messages = {
+        [1] = {
+          role = "system",
+          content = "You are a helpful assistant"
```

Comment:
for a placeholder that must be replaced, it would make sense to explictly make it break if being used. so we can prevent bug that accidently picked up the placeholder. something like content="This is a broken prompt, throw error if you see so".

Reply:
I've dropped this function now, it wasn't flexible enough.